foundry.md

```diff
@@ -96,6 +96,28 @@ First we have some helpers in K which can:
              ...
          </account>
 ```
+```k
+    rule [call.dealInit]:
+         <k> (. => #asInteger(#range(LM, ARGSTART +Int 4, 32)) ) ~> CALL _ CHEAT_ADDR 0 ARGSTART _ARGWIDTH 0 0 ... </k>
+         <output> _ => .ByteArray </output>
+         <localMem> LM </localMem>
+      requires CHEAT_ADDR ==Int #address(FoundryCheat)
+       andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 3364511341 // selector ( "deal" )
+      [priority(40)]
+
+    rule [call.dealFinal]:
+         <k> ACCT_ID:Int ~> CALL _ CHEAT_ADDR 0 ARGSTART _ARGWIDTH 0 0 => 1 ~> #push ... </k>
+         <output> _ => .ByteArray </output>
+         <localMem> LM </localMem>
+         <account>
+           <acctID> ACCT_ID </acctID>
+           <balance> _ => #asWord(#range(LM, ARGSTART +Int 36, 32)) </balance>
+           ...
+         </account>
+      requires CHEAT_ADDR ==Int #address(FoundryCheat)
+       andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 3364511341 // selector ( "deal" )
+      [priority(40)]
```

Comment:
Why are these in two separate rules+ I think they can be just one rule that directly sets `<balance>`?

Reply:
Well, to have them in a single rule I would need to find the account ID by having something like
```
 <acctID> ACCTID </acctID>
...
requires ACCTID ==Int #asInteger(#range(LM, ARGSTART +Int 4, 32))
```

which afaik is not very efficient; and I cannot use directly `<acctID>#asInteger(#range(LM, ARGSTART +Int 4, 32)) </acctID>` because it's a lhs.

Reply:
Did you try putting that on the LHS? Hmmm, I think for concrete execution it should be smart enough to first compute that key, then try finding the map element. For symbolic execution it may be faster to do have you have done.

Reply:
I would still do something like this (single rule for bypassing `CALL`, separate rule for handling `#setBalance`):

```suggestion
    rule [call.deal]:
         <k> #asInteger(#range(LM, ARGSTART +Int 4, 32)) ) ~> CALL _ CHEAT_ADDR 0 ARGSTART _ARGWIDTH 0 0 ... </k>
         <output> _ => .ByteArray </output>
         <localMem> LM </localMem>
      requires CHEAT_ADDR ==Int #address(FoundryCheat)
       andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 3364511341 // selector ( "deal" )
      [priority(40)]

    rule [call.dealFinal]:
         <k> CALL _ CHEAT_ADDR 0 ARGSTART _ARGWIDTH 0 0 => #setBalance(#asWord(#range(LM, ARGSTART +Int 4, 32)), #asWord(#range(LM, ARGSTART +Int 36, 32))) 1 ~> #push ... </k>
         <output> _ => .ByteArray </output>
         <localMem> LM </localMem>
      requires CHEAT_ADDR ==Int #address(FoundryCheat)
       andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 3364511341 // selector ( "deal" )
      [priority(40)]
      
      rule <k> #setBalance(ACCT_ID, NEW_BAL) => . ... </k>
              <account> <acctID> ACCT_ID </acctID> <balance> _ => NEW_BAL </balance> ... </account>
```

Reply:
(obviously my formatting needs to be improved)

Reply:
I will check again with a single rule for `CALL`, but I have doubts since `LM` is not concrete due to the symbolic `value` function arg.

Reply:
Well, the solution above is 2 rules, but more clear than the original solution.